src/select-query-parser/utils.ts

```diff
@@ -0,0 +1,495 @@
+import { FieldNode, Node, SpreadNode } from './parser/ast'
+import {
+  AggregateFunctions,
+  ContainsNull,
+  GenericRelationship,
+  GenericSchema,
+  GenericTable,
+  IsNonEmptyArray,
+  TablesAndViews,
+  UnionToArray,
+} from './types'
+
+export type SelectQueryError<Message extends string> = { error: true } & Message
+
+export type ProcessedNodeResult = Record<string, unknown> | SelectQueryError<string>
+
+type RequireHintingSelectQueryError<
+  DistantName extends string,
+  RelationName extends string
+> = SelectQueryError<`Could not embed because more than one relationship was found for '${DistantName}' and '${RelationName}' you need to hint the column with ${DistantName}!<columnName> ?`>
+
+export type GetFieldNodeResultName<Field extends FieldNode> = Field['alias'] extends string
+  ? Field['alias']
+  : Field['aggregateFunction'] extends AggregateFunctions
+  ? Field['aggregateFunction']
+  : Field['name']
+
+type FilterRelationNodes<Nodes extends Node[]> = UnionToArray<
+  {
+    [K in keyof Nodes]: Nodes[K] extends SpreadNode
+      ? Nodes[K]['target']
+      : Nodes[K] extends FieldNode
+      ? IsNonEmptyArray<Nodes[K]['children']> extends true
+        ? Nodes[K]
+        : never
+      : never
+  }[number]
+>
+
+/*
+ ** Because of pg-meta types generation there is some cases where a same relationship can be duplicated
+ ** if the relation is across schemas and views this ensure that we dedup those relations and treat them
+ ** as postgrest would.
+ ** TODO: This is necessary to allow the check for required desambiguation but shouldn't if the
+ ** generated relationships were correct pointing out that each realation is actually toward a different schema
+ */
+export type DeduplicateRelationships<T extends readonly unknown[]> = T extends readonly [
+  infer First,
+  ...infer Rest
+]
+  ? First extends Rest[number]
+    ? DeduplicateRelationships<Rest extends readonly unknown[] ? Rest : []>
+    : [First, ...DeduplicateRelationships<Rest extends readonly unknown[] ? Rest : []>]
+  : T
+
```

Comment:
**remark**

This should probably be actually handled at `pg-meta` level to avoid duplicating relationships that have the exact same content. Either by providing the `referencedSchema` value or deduplicating the actual keys.

But I think this is actually a much deeper issue about how we handle things, right now everything is focused over a single "schema" so "cross-schema" relations won't work type-wise. I'm not sure but I think it would work at postgrest runtime tough.

I'm pretty confident that this could be implemented at type level as well tough. But it would require changes over the generated `pg-meta` types. So for now, I decided to go to the path of less frictions by simply deduping the relationships to avoid "required hinting errors" to be prompted in such case. But still have types warning where postgrest will actually ask for disambiguation at runtime.

Reply:
Reverted this change since as you said this should be handled at the typegen level. This will be fixed in https://github.com/supabase/postgres-meta/pull/809.

> But I think this is actually a much deeper issue about how we handle things, right now everything is focused over a single "schema" so "cross-schema" relations won't work type-wise

PostgREST doesn't detect relationships across schemas, which is why organize the logic within single schemas.

Reply:
👍